File: 3486-count-the-number-of-good-nodes/3486-count-the-number-of-good-nodes.cpp
class Solution {
public:
    int countGoodNodes(vector<vector<int>>& edges) {
        int n = edges.size() + 1;
        vector<vector<int>> adjacencyList(n);
        for (auto& edge : edges) {
            adjacencyList[edge[0]].push_back(edge[1]);
            adjacencyList[edge[1]].push_back(edge[0]);
        }
        
        vector<int> subtreeSizes(n, 0);
        int goodNodesCount = 0;
        computeSubtreeSizes(adjacencyList, subtreeSizes, goodNodesCount, 0, -1);
        
        return goodNodesCount;
    }
    
private:
    int computeSubtreeSizes(vector<vector<int>>& adjacencyList, vector<int>& subtreeSizes, int& goodNodesCount, int currentNode, int parentNode) {
        int currentSize = 1;
        unordered_map<int, int> sizeFrequency;
        
        for (int neighbor : adjacencyList[currentNode]) {
            if (neighbor == parentNode) continue;
            int childSubtreeSize = computeSubtreeSizes(adjacencyList, subtreeSizes, goodNodesCount, neighbor, currentNode);
            currentSize += childSubtreeSize;
            sizeFrequency[childSubtreeSize]++;
        }
        
        if (sizeFrequency.size() <= 1) goodNodesCount++;
        subtreeSizes[currentNode] = currentSize;
        return currentSize;
    }

    void buildGraph(const vector<vector<int>>& edges, vector<vector<int>>& adjacencyList) {
        int n = edges.size() + 1;
        adjacencyList.resize(n);
        for (const auto& edge : edges) {
            adjacencyList[edge[0]].push_back(edge[1]);
            adjacencyList[edge[1]].push_back(edge[0]);
        }
    }
};

File: 3486-count-the-number-of-good-nodes/3486-count-the-number-of-good-nodes.py
class Solution:
    def countGoodNodes(self, edges: List[List[int]]) -> int:
        tree = defaultdict(list)
        for i, j in edges:
            tree[i].append(j)
            tree[j].append(i)      
        def dfs(node, parent):
            s = 1
            childs = []
            for i in tree[node]:
                if i == parent:
                    continue
                size = dfs(i, node)
                s += size
                childs.append(size)
            if len(set(childs)) <= 1:
                self.good_nodes += 1
            return s
        self.good_nodes = 0
        dfs(0, -1)
        return self.good_nodes